main/main.cpp

```diff
@@ -1,445 +1,382 @@
-
-        case '@':
-            return from_elements({MorseElement::Dot, MorseElement::Dash, MorseElement::Dash, MorseElement::Dot, MorseElement::Dash, MorseElement::Dot}, word_end);
-
-        default:
-            return {{MorseElement::MarkSep}};  // PH error
-        }
-        // clang-format on
-    }
-};
+        // Use PID on the error and add a feed forward term which assumes constant acceleration motion (trapezoid motion
+        // profile).
+        return meters_per_second{
+            std::copysignf(kv, err) * std::sqrt(2 * Motor::max_acceleration * std::abs(err)) + pid.calculate_pid(err)
+        };
+    };
+    const auto limit_velocity = [&](float new_velocity, float last_velocity)
+    {
+        static constexpr auto max_accelaration =
+            Motor::max_acceleration
+            * duration_cast<std::chrono::duration<float, std::ratio<1>>>(pid_loop_period_us).count();
+        return PhysicalMotorSpeed{
+            std::clamp(new_velocity, last_velocity - max_accelaration, last_velocity + max_accelaration)
+        };
+    };
+    const auto calc_motor_velocity = [&](MotorArgs &motor, bool left, float dist_err, float angle_err)
+    {
+        const LinearMotorSpeed wanted_linear_velocity{
+            calc_wanted_velocity(motor.linear_velocity_distance_pid, dist_err, motor.distance_linear_Kv).count()
+        };
+        const RotationalMotorSpeed wanted_angular_velocity{
+            calc_wanted_velocity(motor.angular_velocity_angle_pid, angle_err, motor.distance_angular_Kv).count()
+        };
+
+        motor.wanted_velocity = limit_velocity(
+            left ? wanted_linear_velocity.get() + wanted_angular_velocity.get()
+                 : wanted_linear_velocity.get() - wanted_angular_velocity.get(),
+            motor.wanted_velocity.count().get()
+        );
+        return std::copysignf(motor.Ks, motor.wanted_velocity.count())
+             + motor.velocity_Kv * motor.wanted_velocity.count()
+             + motor.velocity_pid.calculate_pid(motor.wanted_velocity.count() - motor.current_velocity.count());
+    };
+    const MotorSpeed left_motor_velocity{calc_motor_velocity(pid_args->left, true, dist_err, angle_err)};
+    const MotorSpeed right_motor_velocity{calc_motor_velocity(pid_args->right, false, dist_err, angle_err)};
+
+    // Outputs:
+    pid_args->left.output = Motor::bdc_mcpwm_duty_tick_max * left_motor_velocity / Motor::max_speed;
+    pid_args->right.output = Motor::bdc_mcpwm_duty_tick_max * right_motor_velocity / Motor::max_speed;
+    pid_args->left.motor.set_pwm(pid_args->left.output);
+    pid_args->right.motor.set_pwm(pid_args->right.output);
+}
 
-constexpr bool is_word_end(char c)
+static void print_log(const PidArgs &args, unsigned long cycle_time) noexcept
 {
-    switch (c)
-    {
-    case ' ':
-    case '\n':
-    case '\r':
-    case '\t':
-    case '\v':
-    case '\0':
-        return true;
-
-    default:
-        return false;
-    }
+    printf(
```

Comment:
Align with other C library calls:

```suggestion
    std::printf(
```